OpenShift on VMware Cloud on AWS

jcpowermac opened the pull request:
Reuses existing AWS IPI terraform and vSphere UPI
terraform to support VMware Cloud on AWS which
will be the replacement of Packet for CI.

jcpowermac commented:
This most certainly will not pass CI.  Would like to start getting review while I am still testing and modifying.
Needs documentation updates as well.

DanyC97 commented:
@jcpowermac just a heads up in case you haven't seen https://github.com/openshift/installer/pull/2057 which i was trying to get it in for a long time but it was put on hold by the installer team lead.

While i understand from your work that you are focusing on extending the work done initially to run the whole VMware deployment in AWS, i do worry that you are changing the scope which will affect folks who were using this upi sample to bootstrap OCP on a standalone VMware env.

If the installer team desire is to have this upi examples used for CI only then may i suggest you add a bold note so everyone knows that they are not consumable outside RH CI ?

jcpowermac commented:
> @jcpowermac just a heads up in case you haven't seen #2057 which i was trying to get it in for a long time but it was put on hold by the installer team lead.
> 
> While i understand from your work that you are focusing on extending the work done initially to run the whole VMware deployment in AWS, i do worry that you are changing the scope which will affect folks who were using this upi sample to bootstrap OCP on a standalone VMware env.
> 
> If the installer team desire is to have this upi examples used for CI only then may i suggest you add a bold note so everyone knows that they are not consumable outside RH CI ?

Hi @DanyC97 if you have any suggestions I am certainly open to it.  Though atm my priority is to have CI green which is why we are moving to VMC.  I teased apart the IPAM and Route53 terraform from the virtual machine creation so that someone could actually use it without requiring either. That is how I initially tested VMC: ran the AWS terraform that created DNS and LB then ran the terraform that created the virtual machines.  Theoretically as long as you had DNS configured correctly with whatever you use it _should_ work.

patrickdillon commented on upi/vsphere/modules/aws/worker/main.tf:
```diff
@@ -0,0 +1,11 @@
+locals {
+  arn = "aws"
```
is this being used?

patrickdillon commented on upi/vsphere/modules/aws/master/main.tf:
```diff
@@ -0,0 +1,11 @@
+locals {
+  arn = "aws"
```
is this being used?

wking commented on upi/vsphere/README.md:
```diff
@@ -1,59 +1,47 @@
 
-* terraform
-* jq
+The terraform examples provided require a AWS account.  AWS is used
```
nit: "terraform" -> "Terraform"

wking commented on upi/vsphere/README.md:
```diff
@@ -1,59 +1,47 @@
+The `./modules/rhcos_virtual_machines` can be used independently if so desired.
 
-# Build a Cluster
+We will begin with using `openshift-installer` as each terraform example 
```
nit: `openshift-installer` -> `openshift-install`

wking commented on upi/vsphere/README.md:
```diff
@@ -1,59 +1,47 @@
 
-# Build a Cluster
+We will begin with using `openshift-installer` as each terraform example 
+starts with creating the ignition files.
```
nit: "ignition" -> "Ignition", here and later.

wking commented on upi/vsphere/README.md:
```diff
@@ -1,59 +1,47 @@
 
-1. Create an install-config.yaml.
-The machine CIDR for the dev cluster is 139.178.89.192/26.
+## Create ignition files for OpenShift RHCOS virtual machines
```
nit: We can probably drop "for OpenShift RHCOS virtual machines" while remaining sufficiently specific.  For comparison, AWS and GCP docs have "Create Ignition configs" (they disagree on Configs vs. configs, I don't think we have a clear title vs. sentence case policy for h2+ ;).

wking commented on upi/vsphere/README.md:
```diff
@@ -1,59 +1,47 @@
-    defaultDatastore: nvme-ds1
-pullSecret: YOUR_PULL_SECRET
-sshKey: YOUR_SSH_KEY
+cd ./upi/vsphere/
```
nit: no need for the leading `./` or trailing `/`.  Not that it matters much, for folks copy/pasting.

wking commented on upi/vsphere/README.md:
```diff
@@ -1,59 +1,47 @@
-pullSecret: YOUR_PULL_SECRET
-sshKey: YOUR_SSH_KEY
+cd ./upi/vsphere/
+cp install-config.yaml{.example,}
```
nit: brace expansion is [a Bashism][1].  We'd be portable across all POSIX shells with:

```sh
cp install-config.yaml.example install-config.yaml
```

[1]: https://mywiki.wooledge.org/Bashism

wking commented on upi/vsphere/README.md:
````diff
@@ -1,59 +1,47 @@
+
+You can create a `install-config.yaml` file by copying the example provided:
 
 ```
````
nit: doesn't really add much in the way of syntax highlighting in most cases, but it's nice to explicitly declare this to be a shell block with ```sh.

wking commented on upi/vsphere/install-config.yaml.example:
```diff
@@ -0,0 +1,16 @@
+---
```
nit: no need for this leading document separator, and most of our other YAML examples omit it.

wking commented on upi/vsphere/install-config.yaml.example:
```diff
@@ -0,0 +1,16 @@
+    datacenter:
+    defaultDatastore:
+pullSecret:
+sshKey:
```
nit: did we want to fill these in with placeholders/examples like [these examples][1]?

[1]: https://github.com/openshift/installer/blob/8e75675f373d7ca46ec7fa4b4156ce75287735d1/docs/user/customization.md#additional-trust-bundle

wking commented on upi/vsphere/modules/aws/README.md:
```diff
@@ -0,0 +1,4 @@
+# VMware Cloud on AWS
+This is an internal terraform module only.  
+Not particularly useful for anything other than installing
+OpenShift on VMware Cloud on AWS
```
nit: maybe rephrase to:

> This is an internal Terraform module for installing OpenShift in a VMware Cloud on AWS.

That makes implicitly not for other uses, so we don't have to talk about how not-useful it is going to be for those off-label uses ;)

wking commented on upi/vsphere/modules/aws/variables.tf:
```diff
@@ -0,0 +1,95 @@
+  type = "list"
+}
+
+// TODO: this might need to be a list...but may not work right with vsphere on aws
```
Do you want to address this before merging?  I don't see a problem with a single zone for Terraform modules we use for internal testing.  Folks who want a vSphere cluster are probably not going to be using AWS, because if you want a cluster on AWS you'd just use our native AWS providers, right?

wking commented on upi/vsphere/modules/aws/vpc/common.tf:
```diff
@@ -0,0 +1,24 @@
+// Only reference data sources which are guaranteed to exist at any time (above) in this locals{} block
+locals {
+  // How many AZs to create subnets in
+  new_az_count = "${length(var.availability_zones)}"
```
you can drop this if you go with a single zone.

wking commented on upi/vsphere/modules/rhcos_virtual_machines/main.tf:
```diff
@@ -28,74 +28,58 @@ module "bootstrap" {
 }
 
-module "control_plane" {
+module "master" {
```
I know the AWS Terraform uses master/worker, but if you already have control-plane/compute in the vSphere Terraform I'd rather stay with it instead of backsliding.  See also #1330.

wking commented on upi/vsphere/packet/README.md:
````diff
@@ -0,0 +1,106 @@
+cd ./upi/vsphere/packet
+cp terraform.tfvars{.example,}
+```
+Then modify terraform.tfvars file based your environment.
````
nit: backticks around paths and other literals: `terraform.tfvars`.

wking commented on upi/vsphere/vmc/README.md:
````diff
@@ -0,0 +1,130 @@
+openshift-install wait-for install-complete
+```
+When the command completes it will provide authentication details and the URL to the OpenShift console.
+
````
nit: trailing empty line.

jcpowermac commented on upi/vsphere/modules/aws/vpc/outputs.tf:
```diff
@@ -0,0 +1,29 @@
+output "aws_lb_target_group_arns_length" {
+  // 2 for private endpoints and 1 for public endpoints
+  // 2 ingress endpoints
+  value = "5"
```
@patrickdillon /shrug I took the AWS terraform from aws IPI and modified it for VMC. Even in 4.3 its a hard coded output:

https://github.com/openshift/installer/blob/3b221a92063e75883e8872325c19a07bfae1c723/data/data/aws/vpc/outputs.tf#L48

jcpowermac commented:
Thanks @wking and @patrickdillon for taking the time to review.  I am in process of cleaning it up.  Will need to retest.

jcpowermac commented:
@wking and @patrickdillon I think I addressed all the concerns.  Though I wonder if we shouldn't move to terraform 0.12?  From the [template](https://github.com/openshift/release/blob/master/ci-operator/templates/openshift/installer/cluster-launch-installer-upi-e2e.yaml) it looks like only vSphere is using terraform.  The only other UPI that is still terraform 0.11 is Metal but it doesn't look like that template is used with it.

jcpowermac commented:
/hold
Started having issues with injecting ignition into extra config.

jstuever commented:
/uncc @jstuever

jcpowermac commented:
/close
